Ordered Transform

bwengals opened the pull request:
Transform to constrain a random vector's elements to be nondecreasing.  First commit thanks to @aseyboldt.  I'd also like to include an equivalent to Stan's `ordered_logistic` distribution in this PR.

bwengals committed:
initial commit of Ordered, courtesy @aseyboldt

bwengals committed:
initial commit of Ordered, courtesy @aseyboldt

junpenglao commented on pymc3/distributions/transforms.py:
```diff
@@ -236,6 +237,30 @@ def jacobian_det(self, x):
+        return out
+
+    def forward_val(self, x, point=None):
+        x, = pm.distributions.distribution.draw_values([x], point=point)
```
`draw_values` was already imported from `distribution` earlier in the file.

fonnesbeck commented:
Maybe add WIP to the title if you are still looking to add code.

bwengals commented:
What do you guys think of an API similar to `Bound`? 

    OrderedPrior = pm.Ordered(pm.Flat, 'increasing')

Maybe this makes it more straightforward to compose transformations, and the user doesn't need to specify a `transform` arg on a distribution.

junpenglao commented:
Do you mean something like the `Bound`? which you do:
```python
OrderedPrior = pm.Ordered(pm.Gamma, 'increasing')
with pm.Model():
    ...
    a = OrderedPrior('a', alpha=1., beta=1.)
```
In that case, instead of a composite transform you do this?:
```python
OrderedPrior = pm.Ordered(pm.Gamma, 'increasing')
with pm.Model():
    ...
    a = OrderedPrior('a', alpha=1., beta=1., 
                                 transform=pm.distributions.transforms.LogExpM1())
```

bwengals commented:
Yes, that's what I meant.  Although it's implemented as a transform, it's an additional constraint on a prior in the same way `Bound` is.  Then you can handle composing multiple transforms within the `Ordered` object and the composite transform isn't directly exposed to users.

junpenglao commented:
I like your idea. Would it be better to put it in the same file as `Bound`?

bwengals committed:
add composed transform (from @aseyboldt)

bwengals committed:
add composed transform (from @aseyboldt)

twiecki commented:
Ping @bwengals.

junpenglao commented:
FYI: the current version does not work for 2d array. I did some edit here: https://discourse.pymc.io/t/order-statistics-in-pymc3/617/3?u=junpenglao
Not sure if we can make it work for both 1d and 2d tho...

bwengals committed:
add ordered logistic distr

bwengals committed:
add ordered logistic distr

bwengals committed:
add test

bwengals committed:
add test

bwengals committed:
change from function to subclass of Categorical

bwengals committed:
change from function to subclass of Categorical

bwengals commented:
Whoops just saw the ping.  I started looking at this again the last couple days, been working on the OrderedLogistic distribution part.  

I'm been having a hard time getting the ordered transform to work like `Bound` though.  Would it be alright to leave them in just as transforms?

Not sure about the 1D vs 2D transform thing, will look at that

bwengals committed:
test passing on OrderedLogistic

bwengals committed:
test passing on OrderedLogistic

junpenglao commented:
> I'm been having a hard time getting the ordered transform to work like Bound though. Would it be alright to leave them in just as transforms?

Agree. In that case could you please also add `Composed` transforms https://discourse.pymc.io/t/mixture-models-and-breaking-class-symmetry/208/5?u=junpenglao?

bwengals committed:
fix docs up

bwengals committed:
fix docs up

bwengals commented:
Ok awesome :), yup, `Composed` is in here.

bwengals committed:
fix to docstring

bwengals committed:
fix to docstring

bwengals committed:
Merge branch 'master' into ordered

bwengals committed:
Merge branch 'ordered' of github.com:bwengals/pymc3 into ordered

junpenglao commented:
Seems there would be an error in the Jacobian determinant when we are using `Composite` with `Transform` (see discussion here: https://discourse.pymc.io/t/sampling-uniformly-in-a-triangular-support/765/7?u=junpenglao)
cc @aseyboldt

bwengals committed:
starting transform tests

bwengals committed:
starting transform tests

bwengals commented:
sorry for slowness, been having some difficulty getting the tests to work on the ordered/composite transforms.  Just want to status update anyone watching this

bwengals committed:
Merge branch 'ordered' of https://github.com/bwengals/pymc3 into bwengals-ordered

bwengals committed:
Merge branch 'bwengals-ordered' into ordered

bwengals committed:
Merge branch 'master' into ordered

bwengals commented:
I've been stuck on getting these tests to pass :(.  Anyone have any experience with the transforms testing framework?

junpenglao commented on pymc3/distributions/transforms.py:
```diff
@@ -237,6 +238,53 @@ def jacobian_det(self, x):
+class Ordered(ElemwiseTransform):
+    name = "ordered"
+
+    def backward(self, x):
```
Are you sure this is not forward?? See
https://discourse.pymc.io/t/sampling-uniformly-in-a-triangular-support/765/12?u=junpenglao

junpenglao commented on pymc3/tests/test_transforms.py:
```diff
@@ -93,6 +93,17 @@ def check_jacobian_det(transform, domain,
+    check_jacobian_det(tr.ordered, Vector(R, 2),
+                      tt.dvector, np.array([0, 0]), elemwise=True)
+
+def test_ordered_vals():
```
Something like:
```python
def test_ordered_vals():
    vals = get_values(tr.ordered, Vector(R, 3),
                      tt.dvector, np.zeros(3))
    close_to_logical(np.diff(vals) >= 0, True, tol)
```
Again, this test shows that the forward and backward is reversed in your current implementation.

junpenglao commented on pymc3/tests/test_transforms.py:
```diff
@@ -93,6 +93,17 @@ def check_jacobian_det(transform, domain,
+def test_ordered():
+    check_vector_transform_identity(tr.ordered, Vector(R, 6))
+
+def test_ordered_jacobian_det():
```
```python
def test_ordered_jacobian_det():
    check_jacobian_det(tr.ordered, Vector(R, 2),
                      tt.dvector, np.array([0, 0]), elemwise=False)
```

junpenglao commented:
Also need tests for the `Composed` transformation

junpenglao commented on pymc3/tests/test_transforms.py:
```diff
@@ -93,6 +93,17 @@ def check_jacobian_det(transform, domain,
             computed_ljd(yval), tol)
 
+def test_ordered():
+    check_vector_transform_identity(tr.ordered, Vector(R, 6))
```
The domain of ordered transform is not real value, but sorted real value. You should create a new domain `SortedVector` here.

junpenglao replied:
Maybe something like:
```python
# maybe this could be put into test_distributions.py
def SortedVector(n):
    vals = []
    np.random.seed(42)
    for _ in range(10):
        vals.append(np.sort(np.random.randn(n)))
    return Domain(vals, edges=(None, None))

def test_ordered():
    check_vector_transform_identity(tr.ordered, SortedVector(6))
```

bwengals committed:
ordered transform tests pass locally, need tests for composed transform

bwengals committed:
ordered transform tests pass locally, need tests for composed transform

bwengals committed:
Merge branch 'ordered' of github.com:bwengals/pymc3 into ordered

bwengals committed:
Merge branch 'master' into ordered

bwengals commented:
ah thanks a ton @junpenglao, that completely cleared things up for me!

bwengals committed:
fix OrderedLogistic import

bwengals committed:
fix OrderedLogistic import

fonnesbeck commented:
Definitely add this to the change log before merging.

twiecki commented:
@bwengals Is this finished? Would be great to merge (needs rebase and release-notes).

junpenglao commented:
Need to add tests for the `Composed` transformation and release note.

ColCarroll commented on pymc3/tests/test_distributions.py:
```diff
@@ -329,11 +338,18 @@ def mvt_logpdf(value, nu, Sigma, mu=0):
     return (sp.norm(loc=0,scale=1/np.sqrt(tau_e)).logpdf(value[0]) +
             sp.norm(loc=k*value[:-1],scale=1/np.sqrt(tau_e)).logpdf(value[1:]).sum())
 
+def invlogit(x, eps=sys.float_info.epsilon):
```
`sys.float_info.epsilon`: neat!

junpenglao commented:
@bwengals any update?

bwengals commented:
last I picked it up, got stuck on getting the jacobian tests for the `Composed` transform working.  Everything else is good to go, included the `OrderedLogistic` likelihood.  Been seeing what you guys have in this https://discourse.pymc.io/t/sampling-uniformly-in-a-triangular-support/765/14

I suppose we could have `Composed` be a separate PR, and open an issue for it in the meantime.  I can take another crack at it this weekend too and see.  What do you think?

junpenglao commented:
No problem! Also you can have a look at the composite transformation in tensorflow: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/distributions/python/ops/bijectors/chain.py

bwengals committed:
remove composed

bwengals committed:
remove composed

bwengals committed:
a little cleanup

bwengals committed:
a little cleanup

bwengals committed:
Merge branch 'master' into ordered

mkcor committed:
Remove extra the's (#2960)

Commit:
fix test, remove Composed Transform)

Commit:
fix test, remove Composed Transform)

Commit:
Add release note

Commit:
Add release note

junpenglao commented:
@bwengals hope you dont mind I go ahead and fixed the test (need to use the ordered transformation soonish myself ;-) )

Commit:
Ordered transformed is not elementwise

Commit:
Ordered transformed is not elementwise

junpenglao commented:
Please do not merge yet - I think there is some potential problems of the Jacobian.

bwengals commented:
Thanks for bailing me out!

gBokiau committed:
Fix Binomial when p=0 and x=0 (#2961)

* Fix Binomial when p=0 and x=0

Should return 0 not -inf

* Fix logpow(x=0,m=0) 

Fix logpow(x=0,m=0) where x and m have different shapes.

Commit:
Fix Jacobian det

- add test
- add namespace
- fix docstring example

Commit:
Fix Jacobian det

- add test
- add namespace
- fix docstring example

Commit:
Extend to multiple dimension array version

Commit:
Extend to multiple dimension array version

Commit:
Add example

Commit:
Add example

Commit:
PEP8 minor fix

Commit:
PEP8 minor fix

junpenglao commented:
OK this is now ready for review.

twiecki commented:
Also needs a rebase.

junpenglao commented:
Can I even do that? This is not my branch.
[EDIT] Done.

Commit:
rebase

Commit:
clearn up rebase error

twiecki commented on pymc3/distributions/transforms.py:
```diff
@@ -238,6 +242,33 @@ def jacobian_det(self, x):
+
+    def backward(self, y):
+        x = tt.zeros(y.shape)
+        x = tt.inc_subtensor(x[..., 0], y[..., 0])
```
This looks trivial to have figured out :P.

junpenglao replied:
I know - it blows my mind when I came across this.

Commit:
Bug fix

Also extend SumTo1 so that it works in multi-dimension

Commit:
dimension fix

make sure the return jacobian_det has the right dimension: 
- if the transformation is elementwise, return jacob_det should has the same dimension as the input
- if the transformation is on a vector, return jecob_det should has one less dimension

junpenglao commented:
I did some additional tests and it should be fine for this PR now.

I will submit another PR for the chain transformation and add some additional tests.

twiecki commented:
Congrats @junpenglao and @bwengals, quite a tricky one!
